File: elementary_data_structures/stacks.py
class Element:

    __slots__ = ['data', 'next']

    def __init__(self, data):
        self.data = data
        self.next = None


class StackEmptyException(Exception):

    pass


class Stack:

    __slots__ = ['size', 'head']

    def __init__(self):
        self.size = 0
        self.head = None

    def push(self, z):
        self.size += 1

        z.next = self.head
        self.head = z

    def pop(self):
        if self.size == 0:
            raise StackEmptyException

        self.size -= 1

        popped = self.head
        result = popped.data
        self.head = popped.next
        popped = None

        return result

    def print_stack(self):
        print(f"size = {self.size} : ", end='')

        p = self.head
        while p:
            print(f"{p.data}", end=' -> ' if p.next else '\n')

            p = p.next


if __name__ == '__main__':
    s = Stack()

    while True:
        print("1:push 2:pop 3:print > ", end='')
        op = int(input())

        if op == 1:
            print("input data > ", end='')
            data = int(input())

            s.push(Element(data))
        elif op == 2:
            try:
                result = s.pop()

                print(f"result = {result}")
            except StackEmptyException:
                print("stack is empty.")
        elif op == 3:
            s.print_stack()
        else:
            print("invalid operation")

File: elementary_data_structures/stacks.go
//go:build ignore

package main

import "fmt"

type Element struct {
	data int
	next *Element
}

type Stack struct {
	size int
	head *Element
}

func (s *Stack) push(z *Element) {
	s.size += 1

	z.next = s.head
	s.head = z
}

func (s *Stack) pop() int {
	if s.size == 0 {
		fmt.Println("stack is empty.")

		return -1
	}

	s.size -= 1

	e := s.head
	result := e.data
	s.head = e.next
	e = nil

	return result
}

func (s *Stack) printStack() {
	fmt.Printf("size = %d : ", s.size)

	p := s.head
	for p != nil {
		fmt.Printf("%d", p.data)
		if p.next != nil {
			fmt.Print(" -> ")
		}

		p = p.next
	}

	fmt.Println()
}

func main() {
	s := Stack{0, nil}

	for {
		fmt.Print("1:push 2:pop 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)

			z := Element{data, nil}
			s.push(&z)
		case 2:
			result := s.pop()
			if result != -1 {
				fmt.Printf("result = %d\n", result)
			}
		case 3:
			s.printStack()
		default:
			fmt.Println("invalid operation")
		}
	}
}
